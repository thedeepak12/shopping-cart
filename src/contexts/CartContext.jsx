import { createContext, useState, useContext, useEffect } from 'react';

const CartContext = createContext();

export function useCart() {
  return useContext(CartContext);
}

export function CartProvider({ children }) {
  const [cartItems, setCartItems] = useState([]);
  const [cartCount, setCartCount] = useState(0);
  const [cartTotal, setCartTotal] = useState(0);

  useEffect(() => {
    const storedCart = localStorage.getItem('cart');
    if (storedCart) {
      setCartItems(JSON.parse(storedCart));
    }
  }, []);

  useEffect(() => {
    localStorage.setItem('cart', JSON.stringify(cartItems));

    const count = cartItems.reduce((total, item) => total + item.quantity, 0);
    const price = cartItems.reduce(
      (total, item) => total + item.price * item.quantity,
      0
    );

    setCartCount(count);
    setCartTotal(price);
  }, [cartItems]);

  const addToCart = (product) => {
    setCartItems((prevItems) => {
      const existingItemIndex = prevItems.findIndex(
        (item) => item.id === product.id
      );

      if (existingItemIndex >= 0) {
        const updatedItems = [...prevItems];
        updatedItems[existingItemIndex] = {
          ...updatedItems[existingItemIndex],
          quantity: updatedItems[existingItemIndex].quantity + 1,
        };
        return updatedItems;
      }
      return [...prevItems, { ...product, quantity: 1 }];
    });
  };

  const removeFromCart = (productId) => {
    setCartItems((prevItems) =>
      prevItems.filter((item) => item.id !== productId)
    );
  };

  const updateQuantity = (productId, quantity) => {
    if (quantity <= 0) {
      removeFromCart(productId);
      return;
    }

    setCartItems((prevItems) =>
      prevItems.map((item) =>
        item.id === productId ? { ...item, quantity } : item
      )
    );
  };

  const value = {
    cartItems,
    cartCount,
    cartTotal,
    addToCart,
    removeFromCart,
    updateQuantity,
  };

  return <CartContext.Provider value={value}>{children}</CartContext.Provider>;
}
