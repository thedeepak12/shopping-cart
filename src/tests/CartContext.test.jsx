import { describe, it, expect } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { CartProvider, useCart } from '../contexts/CartContext';

function TestCart() {
  const { cartItems, addToCart } = useCart();

  return (
    <div>
      <p data-testid="cart-count">Items in cart: {cartItems.length}</p>
      <button
        onClick={() => addToCart({ id: 1, name: 'Test Product', price: 10 })}
        data-testid="add-button"
      >
        Add to Cart
      </button>
    </div>
  );
}

describe('Shopping Cart', () => {
  it('should match snapshot', () => {
    const { container } = render(
      <CartProvider>
        <TestCart />
      </CartProvider>
    );
    expect(container).toMatchSnapshot();
  });

  it('should start with an empty cart', () => {
    render(
      <CartProvider>
        <TestCart />
      </CartProvider>
    );

    const cartCount = screen.getByTestId('cart-count');
    expect(cartCount.textContent).toBe('Items in cart: 0');
  });

  it('should add an item to cart', () => {
    render(
      <CartProvider>
        <TestCart />
      </CartProvider>
    );

    const addButton = screen.getByTestId('add-button');
    fireEvent.click(addButton);

    const cartCount = screen.getByTestId('cart-count');
    expect(cartCount.textContent).toBe('Items in cart: 1');
  });
});
