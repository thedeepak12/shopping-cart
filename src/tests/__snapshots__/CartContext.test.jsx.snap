// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Shopping Cart > should match snapshot 1`] = `
<div>
  <div>
    <p
      data-testid="cart-count"
    >
      Items in cart: 
      0
    </p>
    <button
      data-testid="add-button"
    >
      Add to Cart
    </button>
  </div>
</div>
`;
